sidechain/consensus/aura/src/block_importer.rs

```diff
@@ -104,54 +103,48 @@ impl<
 	StateHandler: HandleState<StateT = SgxExternalities>,
 	StateKeyRepository: AccessKey,
 	<StateKeyRepository as AccessKey>::KeyType: StateCrypto,
-	TopPoolExecutor:
-		TopPoolCallOperator<ParentchainBlock, SignedSidechainBlock> + Send + Sync + 'static,
+	TopPoolAuthor: AuthorApi<H256, H256> + OnBlockImported<Hash = H256>,
 	ParentchainBlockImporter:
 		TriggerParentchainBlockImport<SignedParentchainBlock<ParentchainBlock>> + Send + Sync,
 {
 	pub fn new(
 		state_handler: Arc<StateHandler>,
 		state_key_repository: Arc<StateKeyRepository>,
-		top_pool_executor: Arc<TopPoolExecutor>,
+		top_pool_author: Arc<TopPoolAuthor>,
 		parentchain_block_importer: Arc<ParentchainBlockImporter>,
 		ocall_api: Arc<OCallApi>,
 	) -> Self {
 		Self {
 			state_handler,
 			state_key_repository,
-			top_pool_executor,
+			top_pool_author,
 			parentchain_block_importer,
 			ocall_api,
 			_phantom: Default::default(),
 		}
 	}
 
 	fn update_top_pool(&self, sidechain_block: &SignedSidechainBlock::Block) {
-		// FIXME: we should take the rpc author here directly #547.
-
 		// Notify pool about imported block for status updates of the calls.
-		self.top_pool_executor.on_block_imported(sidechain_block);
+		self.top_pool_author.on_block_imported(
+			sidechain_block.block_data().signed_top_hashes(),
+			sidechain_block.hash(),
+		);
 
 		// Remove calls from pool.
 		let executed_operations = sidechain_block
 			.block_data()
 			.signed_top_hashes()
 			.iter()
-			.map(|hash| {
-				// Only successfully executed operations are included in a block.
-				ExecutedOperation::success(*hash, TrustedOperationOrHash::Hash(*hash), Vec::new())
-			})
+			.map(|hash| (TrustedOperationOrHash::Hash(*hash), true))
 			.collect();
 
 		let calls_failed_to_remove = self
-			.top_pool_executor
-			.remove_calls_from_pool(&sidechain_block.header().shard_id(), executed_operations);
+			.top_pool_author
+			.remove_calls_from_pool(sidechain_block.header().shard_id(), executed_operations);
```

Comment:
same optimization potential as above, we save O(n) if `remove_calls_from_pool` can take an iterator.